controlplane/rosa/controllers/rosacontrolplane_controller.go

```diff
@@ -0,0 +1,445 @@
+
+// Reconcile will reconcile RosaControlPlane Resources.
+func (r *ROSAControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := logger.FromContext(ctx)
+
+	// Get the control plane instance
+	rosaControlPlane := &rosacontrolplanev1.ROSAControlPlane{}
+	if err := r.Client.Get(ctx, req.NamespacedName, rosaControlPlane); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Get the cluster
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, rosaControlPlane.ObjectMeta)
+	if err != nil {
+		log.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	if capiannotations.IsPaused(cluster, rosaControlPlane) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	rosaScope, err := scope.NewROSAControlPlaneScope(scope.ROSAControlPlaneScopeParams{
+		Client:         r.Client,
+		Cluster:        cluster,
+		ControlPlane:   rosaControlPlane,
+		ControllerName: strings.ToLower(rosaControlPlaneKind),
+	})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create scope: %w", err)
+	}
+
+	// Always close the scope
+	defer func() {
+		if err := rosaScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	if !rosaControlPlane.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Handle deletion reconciliation loop.
+		return r.reconcileDelete(ctx, rosaScope)
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcileNormal(ctx, rosaScope)
+}
+
+func (r *ROSAControlPlaneReconciler) reconcileNormal(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (res ctrl.Result, reterr error) {
+	rosaScope.Info("Reconciling ROSAControlPlane")
+
+	// if !rosaScope.Cluster.Status.InfrastructureReady {
+	//	rosaScope.Info("Cluster infrastructure is not ready yet")
+	//	return ctrl.Result{RequeueAfter: r.WaitInfraPeriod}, nil
+	//}
+	if controllerutil.AddFinalizer(rosaScope.ControlPlane, ROSAControlPlaneFinalizer) {
+		if err := rosaScope.PatchObject(); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Create the cluster:
+	clusterBuilder := cmv1.NewCluster().
+		Name(rosaScope.ControlPlane.Name).
+		MultiAZ(true).
+		Product(
+			cmv1.NewProduct().
+				ID("rosa"),
+		).
+		Region(
+			cmv1.NewCloudRegion().
+				ID(*rosaScope.ControlPlane.Spec.Region),
+		).
+		FIPS(false).
+		EtcdEncryption(false).
+		DisableUserWorkloadMonitoring(true).
+		Version(
+			cmv1.NewVersion().
+				ID(*rosaScope.ControlPlane.Spec.Version).
+				ChannelGroup("stable"),
+		).
+		ExpirationTimestamp(time.Now().Add(1 * time.Hour)).
+		Hypershift(cmv1.NewHypershift().Enabled(true))
+
+	networkBuilder := cmv1.NewNetwork()
+	networkBuilder = networkBuilder.Type("OVNKubernetes")
```

Comment:
Are there any built-in type to use instead?

Reply:
I don't think there is.

Reply:
![shame-icegif-1](https://github.com/kubernetes-sigs/cluster-api-provider-aws/assets/182850/0e3adff3-cc8c-4298-bcd4-3989f054ca42)